feat(performance): adds queues instrumentation docs for js (#10085)
getsentry/sentry-docs: docs/platforms/javascript/common/performance/instrumentation/queues-instrumentation.mdx

---
title: Queues Instrumentation

> description: "Learn how to capture performance data on your messaging queues."

Reviewer: ```suggestion
description: "Learn how to make sure your messaging queues are performing as expected."
```

sidebar_order: 20
---

> To capture performance data on your messaging queues, you need instrument custom spans and transactions around your queue producers and consumers.

Reviewer: ```suggestion
To ensure that you have performance data about your messaging queues, you'll need to instrument custom spans and transactions around your queue producers and consumers.
```

## Producer Instrumentation

> Use the `Sentry.startSpan` function to wrap your queue producer events in order to start capturing performance metrics. Your span `op` must be set to `queue.publish`. Include the following attributes to enrich your producer spans with queue metrics:

Reviewer: ```suggestion
To start capturing performance metrics, use the `Sentry.startSpan` function to wrap your queue producer events. Your span `op` must be set to `queue.publish`. Include the following attributes to enrich your producer spans with queue metrics:
```

| Attribute | Type | Description |
|:--|:--|:--|
| `messaging.message.id ` | string | The message identifier |
| `messaging.destination.name` | string | The queue or topic name |
| `messaging.message.body.size` | int | Size of the message body in bytes |

You must also include trace headers in your message using `spanToTraceHeader` and `spanToBaggageHeader` so that your consumers can continue your trace once your message is picked up.

<Note>

> Your `queue.publish` span must exist as a child within a parent span in order to be recognized as a producer span. If you do not already have a parent producer span, you can start a new one using `Sentry.startSpan`.

Reviewer: ```suggestion
Your `queue.publish` span must exist as a child within a parent span in order to be recognized as a producer span. If you don't already have a parent producer span, you can start a new one using `Sentry.startSpan`.
```

</Note>

```javascript
import { spanToTraceHeader, spanToBaggageHeader } from '@sentry/core';

app.post('/publish', async (req, res) => { // Route handler automatically instruments a parent span
    await Sentry.startSpan(
        {
            name: 'queue_producer',
            op: 'queue.publish',
            attributes: {
                'messaging.message.id': messageId,
                'messaging.destination.name': 'messages',
                'messaging.message.body.size': messageBodySize,
            },
        },
        async (span) => {
            const traceHeader = spanToTraceHeader(span);
            const baggageHeader = spanToBaggageHeader(span);
            await redisClient.lPush('messages', JSON.stringify({
                traceHeader,
                baggageHeader,
                timestamp: Date.now(),
                messageId,
            }));
        },
    );
});
```

## Consumer Instrumentation

> Use the `Sentry.startSpan` function to wrap your queue consumers in order to start capturing performance metrics. Your span `op` must be set to `queue.process`. Include the following attributes to enrich your consumer spans with queue metrics:

Reviewer: ```suggestion
To start capturing performance metrics, use the `Sentry.startSpan` function to wrap your queue consumers. Your span `op` must be set to `queue.process`. Include the following attributes to enrich your consumer spans with queue metrics:
```

| Attribute | Type | Description |
|:--|:--|:--|
| `messaging.message.id ` | string | The message identifier |
| `messaging.destination.name` | string | The queue or topic name |
| `messaging.message.body.size` | number | Size of the message body in bytes |
| `messaging.message.retry.count ` | number | The number of times a message was attempted to be processed |
| `messaging.message.receive.latency ` | number | The time in milliseconds that a message awaited processing in queue |

Use `Sentry.continueTrace` to connect your consumer spans to their associated producer spans:

<Note>

> Your `queue.process` span must exist as a child within a parent span in order to be recognized as a consumer span. If you do not already have a parent consumer span, you can start a new one using `Sentry.startSpan`.

Reviewer: ```suggestion
Your `queue.process` span must exist as a child within a parent span in order to be recognized as a consumer span. If you don't already have a parent consumer span, you can start a new one using `Sentry.startSpan`.
```
